cairo/ethereum/cancun/vm/instructions/memory_instructions.cairo

```diff
@@ -0,0 +1,259 @@
+    if (cast(err3, felt) != 0) {
+        return err3;
+    }
+
+    // OPERATION
+    let memory = evm.value.memory;
+    let (value_data: felt*) = alloc();
+    uint256_to_bytes32(value_data, [value.value]);
+    tempvar value_bytes = Bytes(new BytesStruct(value_data, 32));
+
+    let memory = evm.value.memory;
+    with memory {
+        expand_by(extend_memory.value.expand_by);
+        memory_write(start_position, value_bytes);
+    }
+
+    // PROGRAM COUNTER
+    EvmImpl.set_pc_stack_memory(Uint(evm.value.pc.value + 1), stack, memory);
+    let ok = cast(0, ExceptionalHalt*);
+    return ok;
+}
+
+// @notice Stores a byte to memory
+func mstore8{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, evm: Evm}() -> ExceptionalHalt* {
+    alloc_locals;
+    // STACK
+    let stack = evm.value.stack;
+    with stack {
+        let (start_position, err1) = pop();
+        if (cast(err1, felt) != 0) {
+            return err1;
+        }
+        let (value, err2) = pop();
+        if (cast(err2, felt) != 0) {
+            return err2;
+        }
+    }
+
+    // GAS
+    tempvar mem_access_tuple = new TupleU256U256(
+        new TupleU256U256Struct(start_position, U256(new U256Struct(1, 0)))
+    );
+    tempvar mem_access_list = ListTupleU256U256(new ListTupleU256U256Struct(mem_access_tuple, 1));
+    let extend_memory = calculate_gas_extend_memory(evm.value.memory, mem_access_list);
+
+    // assumed that cost < 2**110 (see calculate_memory_gas_cost)
+    let err3 = charge_gas(Uint(GasConstants.GAS_VERY_LOW + extend_memory.value.cost.value));
+    if (cast(err3, felt) != 0) {
+        return err3;
+    }
+
+    // OPERATION
+    let (data) = alloc();
+    assert bitwise_ptr.x = value.value.low;
+    assert bitwise_ptr.y = 0xFF;
+    assert [data] = bitwise_ptr.x_and_y;
+    let bitwise_ptr = bitwise_ptr + BitwiseBuiltin.SIZE;
+    tempvar value_bytes = Bytes(new BytesStruct(data, 1));
+
+    let memory = evm.value.memory;
+    with memory {
+        expand_by(extend_memory.value.expand_by);
+        memory_write(start_position, value_bytes);
+    }
+
+    // PROGRAM COUNTER
+    EvmImpl.set_pc_stack_memory(Uint(evm.value.pc.value + 1), stack, memory);
+    let ok = cast(0, ExceptionalHalt*);
+    return ok;
+}
+
+// @notice Load word from memory
+func mload{range_check_ptr, evm: Evm}() -> ExceptionalHalt* {
+    alloc_locals;
+    // STACK
+    let stack = evm.value.stack;
+    with stack {
+        let (start_position, err1) = pop();
+        if (cast(err1, felt) != 0) {
+            return err1;
+        }
+    }
+
+    // GAS
+    tempvar mem_access_tuple = new TupleU256U256(
+        new TupleU256U256Struct(start_position, U256(new U256Struct(32, 0)))
+    );
+    tempvar mem_access_list = ListTupleU256U256(new ListTupleU256U256Struct(mem_access_tuple, 1));
+    let extend_memory = calculate_gas_extend_memory(evm.value.memory, mem_access_list);
+
+    // assumed that cost < 2**110 (see calculate_memory_gas_cost)
```

Comment:
what does it mean that it is assumed?
what happens if cost > 2**110?

Reply:
calculate_gas_extend_memory will saturate if size < 2**64
If not mistaken the cost will be < 2**110 
If those are not respected, it might not be sound

Reply:
> what happens if cost > 2**110?

it won't happen because the cost is saturated